### ISSUE TITLE
IPC error during Bus tests

### ISSUE THREAD

**Issue by mwpowellhtx** [nanomsg/nng#150]
```
2017.11.04 17:55:10.830   ERROR Process C:\Users\Michael\AppData\Local\JetBrains\Installations\ReSharperPlatformVs14\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe:14588 exited with code '3':
panic: G:\Source\Spikes\nanomsg\csnng\prototype\repos\nng\src\platform\windows\win_ipc.c: 65: assert err: aio->a_iov[0].iov_len
This message is indicative of a BUG.
Report this at http://github.com/nanomsg/nanomsg
```

Test log:

```
Running protocol tests for address family 'InterProcess'.
Testing using address 'ipc://pipe/325656c3-4e60-4c89-81d4-097e4a20dfa2'.
Given three Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket instances.
  And messages can be delivered.
    And receive times out.
      And Bus1 delivers message to both Bus2 and Bus3.
```

Based on IPC specialized Bus tests:

```C#
namespace Nanomsg2.Sharp.Protocols.Bus
{
    using Xunit.Abstractions;

    public class InterProcessBusTests : BusTests
    {
        protected override SocketAddressFamily Family { get; } = SocketAddressFamily.InterProcess;

        public InterProcessBusTests(ITestOutputHelper @out)
            : base(@out)
        {
        }
    }
}
```

Running under Visual Studio 2015, C# 6.0, xUnit, Windows 7 Professional (x64).

**Comment by mwpowellhtx** [comment]
I'm looking into my Bus unit tests just besides. Unrelated to this error however. Definitely an issue extending into the Pipeline unit tests, for instance.

**Comment by mwpowellhtx** [comment]
Odd. IPC pub/sub appears to work.

IPC pipeline causing the following error:

```
2017.11.04 19:19:38.505   ERROR Process C:\Users\Michael\AppData\Local\JetBrains\Installations\ReSharperPlatformVs14\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe:14468 exited with code '3':
panic: G:\Source\Spikes\nanomsg\csnng\prototype\repos\nng\src\platform\windows\win_ipc.c: 65: assert err: aio->a_iov[0].iov_len
panic: G:\Source\Spikes\nanomsg\csnng\prototype\repos\nng\src\platform\windows\win_ipc.c: 65: assert err: aio->a_iov[0].iov_len
This message is indicative of a BUG.
This message is indicative of a BUG.
Report this at http://github.com/nanomsg/nanomsg
Report this at http://github.com/nanomsg/nanomsg
```

**Comment by mwpowellhtx** [comment]
IPC also working for req/rep, individually.

However, choking when taken in parallel.

```
2017.11.04 19:25:13.345   ERROR Process C:\Users\Michael\AppData\Local\JetBrains\Installations\ReSharperPlatformVs14\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe:9464 exited with code '3'.
```

**Comment by mwpowellhtx** [comment]
Same for Survey. Fine when taken individually. Chokes in parallel.

```
2017.11.04 19:29:40.955   ERROR Process C:\Users\Michael\AppData\Local\JetBrains\Installations\ReSharperPlatformVs14\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe:12584 exited with code '3'.
```

**Comment by gdamore** [comment]
Parallel test instances are going to do evil things due to address conflicts.  I haven't designed the tests to support parallel execution -- sorry.

Assertion failure are evil though, so I do need to get to the bottom of that.

**Comment by mwpowellhtx** [comment]
Understandably so. I'd contend that threaded/concurrent behavior is a desirable attribute, though.

**Comment by gdamore** [comment]
The *tests* can't be run concurrently because they don't know how to allocate separate, non-conflicting global addresses (IPC paths, TCP port numbers, etc.)  Fixing that is *not* a priority... because the enclosing test framework I use *does* run these in parallel and includes support for giving the test instances each their own "range" of port numbers.

Manually doing this is never going to work unless the manual user figures out how to avoid the conflicts. 

And running the multiple instances of the entire test suite together won't work either, because they don't know about each other, and their conflict avoidance schemes won't work in that case.

**Comment by gdamore** [comment]
The underlying *nng* stack *can* be run concurrently.  Its the *tests* that have limits.

**Comment by mwpowellhtx** [comment]
What's the difference? Obviously it cannot or the tests would be passing with flying colors whether isolated to a specific transport or across all transports.

**Comment by gdamore** [comment]
The issue is that you can't have completely different programs using the same resource (such as a named pipe or TCP port) in conflicting ways without stomping on each other.  This isn't a bug in nng.  I've not tried to engineer the *test* suite to be robust against this stupidity because it isn't a priority.

**Comment by gdamore** [comment]
I'd be very keen to see your test code, because I don't think this assertion should be possible, but I may not have validated that the incoming aio if you're using an aio based approach, is valid.

It's really significant that an IOV is being given to the receive logic with a bad length.

**Comment by mwpowellhtx** [comment]
Watch your disparaging comments. Like I said, the same tests run individually are working. But when run in parallel as a group they are not. Nothing about the tests or the addressing strategy changed in between. So the only other thing I walk away with is that NNG is failing to deal with multiple transports concurrently.

**Comment by gdamore** [comment]
You are *not* understanding the issue, clearly.

The problem is analogous to trying to run and test two HTTP servers both on port 80, *simultaneously*.  Its even worse because the servers may be serving up different content, or may be injecting failures intentionally, and so the behavior is "unpredictable".  This IS NOT AN NNG BUG.  It *is* a failure to understand the limitations of the test suite.

**Comment by gdamore** [comment]
I regularly run the tests in parallel in CI.  All the time.  However, doing so *requires* that the test suite uses the TEST_PORT logic in the CMakefiles in the test directory to assign each test it's own range of 10 unique ports.

**Comment by gdamore** [comment]
The real issue here is that nng is triggering an assertion failure -- that shouldn't happen.

What I would like to know is whether this is occurring because of a failure to pre-validate bad input from the user program, or some other condition internally.  I suspect that its the former, but I can't be sure.  If you had reproduced the assertion failure with my test program rather than yours, it would be easier for me to diagnose -- especially if I have a means to reproduce.

**Comment by mwpowellhtx** [comment]
I disagree, and here's the evidence to prove it. In the bus tests across IPv4 and IPv6 only. Failures across the board. Here are my logs:

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.000] Aborted
Running protocol tests for address family 'IPv4'.
Testing using address 'tcp://127.0.0.1:10001'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
```

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.000] Aborted
Running protocol tests for address family 'IPv6'.
Testing using address 'tcp://[::1]:10000'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
```

I ran both together, and they both fell over. You can clearly see the address, both loopbacks, ***DIFFERENT*** ports.

**Comment by mwpowellhtx** [comment]
So, I don't know what you have to do to repro in C. I don't envy you for that. All I can do is report the runner (Task oriented, I think), has a problem with it.

**Comment by gdamore** [comment]
Those port numbers only differ by 1.   Are you sure that port reuse is not a problem -- for example in many of my tests I continue to increment the port number between subsequent test runs.

I'm assuming these are completely separate programs with separate instances.  There are no global resources that nng uses -- other than the ports or paths that are configured into programs via URLs.  So there is no way for them to interfere with each other unless you have got collisions there.

Remember also with TCP that just closing a port doesn't necessarily free it right back up -- there  is the whole FIN_WAIT thing that operating systems do under the hood.

**Comment by gdamore** [comment]
I can build a test program that demonstrates multiple concurrent use if necessary.  I thought the device tests did, but it turns out that they don't.   Supporting multiple concurrent transports (tcp vs inproc vs ipc)  -- both on the same socket and in separate sockets, is a core competency for nng.  Supporting multiple concurrent "protocols" (e.g. pair and pub/sub) can only be done on different sockets, and should be fine.

**Comment by mwpowellhtx** [comment]
It's possible my addressing strategy is the issue, so I will look into that more closely. Contending with a zombie console as well from a previous console test run. Ugh.

**Comment by mwpowellhtx** [comment]
Back after restart. Please explain what you mean by: 1) bad input, and 2) pre-validate. I take to mean, "ensure addresses are unique". We'd been over that once before, for instance, in my C++ wrappers, which I verified.

**Comment by mwpowellhtx** [comment]
Okay, so after a bit of tinkering, I updated my port allocation strategy. Each unique type of test fixture gets its own base port. Effectively port range. So when a port is requested, they are consumed from there.

When I run my IPv4 tests individually, I get this:

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.279] Success
Running protocol tests for address family 'IPv4'.
Testing using address 'tcp://127.0.0.1:10368'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus1 delivers message to both Bus2 and Bus3.
```

```
   BusTests.That_Bus2_delivers_message_to_Bus1_and_Bus3_times_out [0:00.252] Success
Running protocol tests for address family 'IPv4'.
Testing using address 'tcp://127.0.0.1:10369'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus2 delivers message to Bus1, Bus3 times out.
```

And when I run my IPv6 tests I get this:

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.276] Success
Running protocol tests for address family 'IPv6'.
Testing using address 'tcp://[::1]:10352'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus1 delivers message to both Bus2 and Bus3.
```

```
   BusTests.That_Bus2_delivers_message_to_Bus1_and_Bus3_times_out [0:00.249] Success
Running protocol tests for address family 'IPv6'.
Testing using address 'tcp://[::1]:10353'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus2 delivers message to Bus1, Bus3 times out.
```

However, when I run them both concurrently:

![image](https://user-images.githubusercontent.com/1824187/32411191-859cadde-c1ab-11e7-935f-9ff211a93d17.png)

Versus individually:

![image](https://user-images.githubusercontent.com/1824187/32411192-9e216f16-c1ab-11e7-936f-7e5f5f406faa.png)

**Comment by SG7** [comment]
//127.0.0.1:10368 and 'tcp://127.0.0.1:10369  are very close.  Could
you repeat the test with ports being 1000 apart?

**Comment by gdamore** [comment]
So I'm no longer of the opinion that a bad input (bad AIO) could normally cause the assertion failure you're seeing.  In fact, I can't figure out (still) how this assertion error triggers.  It would be amazing if I had a C (or possibly C++) program that reproduced the issue.  I'm starting to suspect stack or heap corruption.  Being able to attach a debugger and capture the error when it happened would be really helpful.

Just changing the bus test program to use IPC did not seem to trigger this failure (on Windows 10 at least).

I'm pretty sure "error code 3" is me calling abort() in my test programs when some test circumstance fails.

Without more detail, there's not a lot I can do -- I've been staring at the code and trying various things to trigger that assertion on my own, and I haven't had any luck.

**Comment by gdamore** [comment]
Note that on many platforms (with Windows it may depend on which Windows version you run), TCP ports are kind of "shared" between v4 and v6.  They may step on each other, or they might not.

**Comment by mwpowellhtx** [comment]
Same result. From the successful individual runs:

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.280] Success
Running protocol tests for address family 'IPv4'.
Testing using address 'tcp://127.0.0.1:10368'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus1 delivers message to both Bus2 and Bus3.
```

```
   BusTests.That_Bus2_delivers_message_to_Bus1_and_Bus3_times_out [0:00.247] Success
Running protocol tests for address family 'IPv4'.
Testing using address 'tcp://127.0.0.1:11368'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus2 delivers message to Bus1, Bus3 times out.
```

```
   BusTests.That_Bus1_delivers_message_to_both_Bus2_and_Bus3 [0:00.288] Success
Running protocol tests for address family 'IPv6'.
Testing using address 'tcp://[::1]:10352'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus1 delivers message to both Bus2 and Bus3.
```

```
   BusTests.That_Bus2_delivers_message_to_Bus1_and_Bus3_times_out [0:00.240] Success
Running protocol tests for address family 'IPv6'.
Testing using address 'tcp://[::1]:11352'.
Given three 'Nanomsg2.Sharp.Protocols.Bus.LatestBusSocket' instances.
  And messages can be delivered.
    And receive times out.
    And Bus2 delivers message to Bus1, Bus3 times out.
```

I'll push my code up to Github tomorrow so you can have a look. The build process is twofold. One step is CMake, which establishes the NNG environment. Accept the defaults, as the Visual Studio solution is expecting a DLL to land in the CMake out of place build, as opposed to an install prefix, per se.

After that, it's all fairly vanilla DllImport stuff, and Task oriented xUnit test runners.

**Comment by mwpowellhtx** [comment]
At this point I don't think it's a port issue as the behavior extends even into non-port oriented protocols, i.e. InProcess and InterProcess (although IPC is probably a somewhat separate issue, methinks).

**Comment by mwpowellhtx** [comment]
I've already provided ample evidence, I am not sharing ports. Each test fixture gets its own base port, incremented by 16 (literally, +=0x10). Then each subsequent request is different by +1000 from the last. It's not a port sharing issue.

What are you doing to maintain the socket address structure? The state with that can get very confused if you aren't watching the transport family very carefully.

**Comment by gdamore** [comment]
"Inproc" has *zero* shared state between processes.  As in *nothing*, although there is a weird little TCP connection that I create per socket (no long lived listener though) in order to support the stupid select() semantic on Windows.  (Because Windows sockets are "special".)

I'm really really not looking forward to trying to debug C# -- I've never worked in that language before.

I think it's going to be faster and easier for me to simply write a C program that demonstrates that the underlying functionality "works".

**Comment by gdamore** [comment]
To "prove" that you're not sharing ports, I need something beyond random code snippets, because I don't know anything about the test execution flow (do stacks rewind for example) or your runtime.  What seems "ample" evidence to you is insufficient for my purposes.

The best way to "prove" this would be to add some printf's into the places that you call into nng, printing the actual addresses to the console.

I say this, because I've already been surprised once today by my own code reusing addresses when I thought it could not be, because of test execution flow and setjmp; state was getting reset in some very unexpected ways.  I had to use printf() to see what was going on.

**Comment by mwpowellhtx** [comment]
Again with the disparaging comments. I don't know what you're talking about with "random code snippets". That is the ***ONE*** place where I issue the next address. You called it into question, so I showed you precisely what I am doing to resolve the issue. Which, now also keyed on the test fixture originating the request.

It's a fair point, and similar to Convey or Catch, whose execution paths are unwound or whose paths are guaranteed to be unique, I've also ensured that there is no shared state between the code. In other words, the test paths are also unique. I've shown that in the reported addresses; whenever the next address is requested for listen/dial pairs, that's what is used.

I've also gone to great lengths to ensure that each Catch/Convey leaf is represented by at least one xUnit Fact (and/or data-driven Theory, where appropriate), which may call into parent contexts in order to establish their prerequisites. It's analogous if not consistent with the Catch/Convey way of doing things, but I'd be happy to explain that in a live discussion if need be.

I'm not expecting you to be an expert C# developer, and I can help you through that. However, the syntax isn't so foreign from C/C++, however, that it should be that alien to you.

My eyes are crossed over this right now, and I don't want to muck it up pulling into Github. So I will sleep on it and get it up ASAP tomorrow fresh.

**Comment by gdamore** [comment]
I've written a test program, and determined that there exists a bug in the Windows IPC code.

It would appear that I made some bad assumptions in the code, because of the way Windows IPC works.  Basically, we have to specially handle some things because Windows named pipes do not do scatter gather, and some of the protocols have a header, and others do not.  If the protocol-specific header is empty (Bus is one of those!) then we wind up having a zero length "scatter/gather" element.  This was not something I was expecting when I wrote that code, but I had forgotten the particular sense of the code.  Note that REQ/REP and similar protocols that *do* have a header probably never see this problem.

I don't think this has anything to do with threading, or not.  It is simply that Windows IPC and certain protocols are not going to work.

I'm working on a fix.
